IPython/core/inputtransformer2.py

```diff
@@ -0,0 +1,556 @@
+"""Input transformer machinery to support IPython special syntax.
+
+This includes the machinery to recognise and transform ``%magic`` commands,
+``!system`` commands, ``help?`` querying, prompt stripping, and so forth.
+"""
+
+# Copyright (c) IPython Development Team.
+# Distributed under the terms of the Modified BSD License.
+
+from codeop import compile_command
+import re
+import tokenize
+from typing import List, Tuple
+import warnings
+
+_indent_re = re.compile(r'^[ \t]+')
+
+def leading_indent(lines):
+    """Remove leading indentation.
+    
+    If the first line starts with a spaces or tabs, the same whitespace will be
+    removed from each following line.
+    """
+    m = _indent_re.match(lines[0])
+    if not m:
+        return lines
+    space = m.group(0)
+    n = len(space)
+    return [l[n:] if l.startswith(space) else l
+            for l in lines]
```

Comment:
I'm just going to write some +1/ok here and there to have a mark for myself as to what I have already reviewed.